### PR TITLE
feat(config/presets): fetchJSONFile handle branchOrTag in Bitbucket Server 

### LINKED ISSUE #12936: Bitbucket Server local presets don't support path or tag settings
### What would you like Renovate to be able to do?

https://docs.renovatebot.com/config-presets/#self-hosted-git: Bitbucket local presets will only work for files that are stored in repo root.

Given a config such as `local>PROJECT/repostitory:path/preset#feature/branch`, renovate will fail with the following query:
https://git.impello.co.uk/rest/api/1.0/projects/PROJECT/repos/repostitory/browse/path.json?limit=20000

Given a config such as `local>PROJECT/repostitory:preset#feature/branch`, renovate will fail with the following query:
https://git.impello.co.uk/rest/api/1.0/projects/PROJECT/repos/repostitory/browse/preset#feature.json?limit=20000

### If you have any ideas on how this should be implemented, please tell us here.

It looks like it gets the string before first `/` in the config and treats that as the file path. The code need to be updated to split the config string correctly.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

Part 1 of Bitbucket Server local presets don't support path or tag settings #12936

fetchJSONFile handle branchOrTag

## Context:

Closes #12936

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/bitbucket-server/index.ts [hunk 1]
```diff
@@ -17,11 +17,16 @@ const http = new BitbucketServerHttp();
...
-  endpoint: string
+  endpoint: string,
+  branchOrTag?: string
 ): Promise<Preset> {
   const [projectKey, repositorySlug] = repo.split('/');
   setBaseUrl(endpoint);
-  const url = `rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/browse/${fileName}?limit=20000`;
+  let url = `rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/browse/${fileName}?limit=20000`;
+  if (branchOrTag) {
+    url += '&at=' + encodeURIComponent(`refs/heads/${branchOrTag}`);
```
Comment (viceice) [inline lib/config/presets/bitbucket-server/index.ts:27]: does it work with git tags too?

Comment (olegkrivtsov) [reply lib/config/presets/bitbucket-server/index.ts]: @viceice I tried to fix this so it should work with branch names and tags as well.

File: lib/config/presets/bitbucket-server/index.ts [hunk 2]
```diff
@@ -17,11 +17,16 @@ const http = new BitbucketServerHttp();
 export async function fetchJSONFile(
   repo: string,
   fileName: string,
-  endpoint: string
+  endpoint: string,
+  packageTag?: string
```
Comment (viceice) [inline lib/config/presets/bitbucket-server/index.ts:21]:
```suggestion
  branchOrTag?: string
```

Should match interface declaration

File: lib/config/presets/bitbucket-server/index.ts [hunk 3]
```diff
@@ -17,11 +17,16 @@ const http = new BitbucketServerHttp();
...
 ): Promise<Preset> {
   const [projectKey, repositorySlug] = repo.split('/');
   setBaseUrl(endpoint);
-  const url = `rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/browse/${fileName}?limit=20000`;
+  let url = `rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/browse/${fileName}?limit=20000`;
+  if (packageTag) {
+    url += '&at=' + encodeURIComponent(packageTag);
```
Comment (viceice) [inline lib/config/presets/bitbucket-server/index.ts:27]:
```suggestion
  if (branchOrTag) {
    url += '&at=' + encodeURIComponent(branchOrTag);
```

**Review by viceice — APPROVED** [review]

**Pull request merged**